### ISSUE TITLE
API is down

### ISSUE THREAD

**Issue by Abdirahiim** [ExpDev07/coronavirus-tracker-api#264]
So I was doing some unit tests for my project when I saw that they failed and I checked out the API to see if everything is working but it seems that it's down

**Comment by Bost** [comment]
This works flawlessly https://github.com/ExpDev07/coronavirus-tracker-api#running--development

And seriously, this API service has been reaching it's capacity, so using your localhost endpoint for unit testing is more than just **nice** towards the others.

**Comment by Abdirahiim** [comment]
I was testing out my API wrapper and since that it uses the API using it for tests is inevitable

**Comment by AafaaqAli** [comment]
yes, it's down... but why so?

**Comment by paolotamag** [comment]
down for me as well during my webinar XD 
luckily i had previously cached data in my dashboard!

**Comment by azeezgaa** [comment]
Anyone has any idea is this going to solved ?

Since i had a demo to my manager on this scheduled this Friday

**Comment by GabrielDS** [comment]
> Anyone has any idea is this going to solved ?
> 
> Since i had a demo to my manager on this scheduled this Friday

I think the problem is the app on Dyno (instances heroku).  I runned without problems locally. 
@Kilo59

**Comment by ibhuiyan17** [comment]
<img width="1255" alt="Screen Shot 2020-04-07 at 1 19 16 PM" src="https://user-images.githubusercontent.com/33792969/78699968-b8603700-78d2-11ea-9e9e-4681aea43cc5.png">

https://www.heroku.com/pricing

Unfortunately scaling up on Heroku isn't cheap but here are some of the options that can be considered. I think the best option for now is running it locally

**Comment by Toxyl** [comment]
Maybe crowdfunding?

**Comment by Kilo59** [comment]
I have another version of it hosted in the US that doesn't get nearly as much traffic.

https://covid-tracker-us.herokuapp.com/
https://github.com/ExpDev07/coronavirus-tracker-api/issues/248#issuecomment-607513106

**Comment by ibhuiyan17** [comment]
Could some sort of load-balancing scheme be implemented that directs traffic to multiple free instances?

**Comment by Toxyl** [comment]
I'm still trying to get a node working on a Ubuntu server 'cause then I can spin up some droplets and put a load balancer in front.

**Comment by azeezgaa** [comment]
> > Anyone has any idea is this going to solved ?
> > Since i had a demo to my manager on this scheduled this Friday
> 
> I think the problem is the app on Dyno (instances heroku). I runned without problems locally.
> @Kilo59

What do you mean run locally,  can you please help on this. currently i am just using the heroko api and getting the response in json format

**Comment by itsamirrezah** [comment]
@azeezgaa actually it's pretty simple, just follow the instruction that mentioned in Readme file.

* Clone repository 
* Install python 3.8
* Install pipenv
* Run `pipenv shell` command.

Checkout installation section: https://github.com/ExpDev07/coronavirus-tracker-api#installation

**Comment by azeezgaa** [comment]
> @azeezgaa actually it's pretty simple, just follow the instruction that mentioned in Readme file.
> 
> * Clone repository
> * Install python 3.8
> * Install pipenv
> * Run `pipenv shell` command.
> 
> Checkout installation section: https://github.com/ExpDev07/coronavirus-tracker-api#installation

But this will work onlly in my pc not in other pc's right and it wont work on mobile devices am i right

**Comment by itsamirrezah** [comment]
@azeezgaa Yes, your right, but for development & testing purposes, running locally is more reliable way.
also I think you can deploy the repository on your own server. (if its allowed)

**Comment by Toxyl** [comment]
It didn't work quite so well for me the first time around when I tried with Debian 10 and Ubuntu 19.10, Ubuntu 18.04.3 did though. Should have looked at the files earlier, then I would have seen the docker-compose.yml xD Now making an Ansible playbook to deploy those.

**Comment by gribok** [comment]
Many platforms supports COVID-19 projects with enterprise accounts for free during pandemic.
See https://github.blog/2020-03-23-open-collaboration-on-covid-19/

Maybe this could push the project to get on landingpage.

**Comment by Kilo59** [comment]
We don't have good error logging setup so I can't really determine what the underlying cause was.
We don't actually get __THAT__ hot, which seems to be anywhere between 200 - 800 requests per minute depending on the time of day.

Here are some metrics from around the time of the outage and then you can see when the app recovered. It could have just been a problem with Heroku's infrastructure. 🤷‍♂ 
## Outage.
![image](https://user-images.githubusercontent.com/13108583/78721926-72b56580-78f6-11ea-9813-a9292848468c.png)

### Recent
![image](https://user-images.githubusercontent.com/13108583/78722069-bc05b500-78f6-11ea-93eb-62bdf6736503.png)

As you can see we are actually handling a lot more requests now than at the time of the outage.

**Comment by Toxyl** [comment]
I've setup two small droplets running the API behind a load balancer, feel free to use them as long as they can handle the load: http://188.166.130.146 
I've also written an Ansible playbook that I can use to add new nodes when needed.

**Comment by Kilo59** [comment]
FYI it's back up.
It seems to have *only* 😬 lasted about an hour.

**Comment by Toxyl** [comment]
![avg_req_duration](https://user-images.githubusercontent.com/11501185/78861307-caf28180-7a34-11ea-95cd-fe5100a2dee4.PNG)

That is from my load balancer which is experiencing hardly any load. Ignore the first half as I was still doing some setup on the nodes. Makes me wonder if the API has memory leaks or a similar issue which could explain the outage. Can you check logs on heroku? Maybe the API crashed after a while of timing out (i.e. the average request duration growing too high).

**Comment by Kilo59** [comment]
@Toxyl no I can't go back and look at the logs more that far back.
I have added the free tier of Timber IO #266, so in the future, I should be able to look back.

I've also seen those log messages, but I don't know enough about that section to know if that's concerning. If it's not actually a problem I do think we should change the log level for that message because it does imply that there's a problem.

In regards to a possible memory leak being the cause. Not saying that's not possible, but if you look at the [screenshot at the time of the outage](https://github.com/ExpDev07/coronavirus-tracker-api/issues/264#issuecomment-610634512), you'll also notice that it recovers on its own, not because of a redeployment.

I need to spend more time looking over the code, but the only clue I have is that we cache our source data results for an hour and the outage lasted for about an hour.
But typically things like Python's LRU cache do not cache exceptions.
That isn't what we are using but it's possible our solution does cache exceptions or the results that cached caused downstream exceptions.
An hour later the cache was refreshed and the application recovered 🤷 .

Going to close this issue since the API has been back up for days at this point, but I will open another issue to continue this investigation.  

https://github.com/ExpDev07/coronavirus-tracker-api/issues/270

**Comment by Toxyl** [comment]
> @Toxyl no I can't go back and look at the logs more that far back.
> I have added the free tier of Timber IO #266, so in the future, I should be able to look back.
> 
> I've also seen those log messages, but I don't know enough about that section to know if that's concerning. If it's not actually a problem I do think we should change the log level for that message because it does imply that there's a problem.
> 
> In regards to a possible memory leak being the cause. Not saying that's not possible, but if you look at the [screenshot at the time of the outage](https://github.com/ExpDev07/coronavirus-tracker-api/issues/264#issuecomment-610634512), you'll also notice that it recovers on its own, not because of a redeployment.
> 
> I need to spend more time looking over the code, but the only clue I have is that we cache our source data results for an hour and the outage lasted for about an hour.
> But typically things like Python's LRU cache do not cache exceptions.
> That isn't what we are using but it's possible our solution does cache exceptions or the results that cached caused downstream exceptions.
> An hour later the cache was refreshed and the application recovered 🤷 .
> 
> Going to close this issue since the API has been back up for days at this point, but I will open another issue to continue this investigation.
> 
> #270

I don't know how you've set up the server. The way I would set it up it would restart the API when it crashes, which would happen after a while of timeouts, so it would appear unreachable for a while before coming back. If it was a memory leak the API would crash after all memory has been consumed which does not imply that it's reachable until then. But yeah, I might be way off with that, it's just a hunch.

**Comment by Bost** [comment]
> with gradually more warnings printed at the same time. By now the warnings look like this:
> 
> ```
> WARNING:  No country code found for 'Diamond Princess'. Using 'XX'!
> WARNING:  No country code found for 'MS Zaandam'. Using 'XX'!
> WARNING:  No country code found for 'Diamond Princess'. Using 'XX'!
> WARNING:  No country code found for 'MS Zaandam'. Using 'XX'!
> WARNING:  No country code found for 'Diamond Princess'. Using 'XX'!
> WARNING:  No country code found for 'Diamond Princess'. Using 'XX'!
> WARNING:  No country code found for 'MS Zaandam'. Using 'XX'!
> WARNING:  No country code found for 'MS Zaandam'. Using 'XX'!
> ```
> 
> I'll keep an eye on this to see if the trend continues.

The API service was looking for a country code in the [COUNTRY_NAME__COUNTRY_CODE](https://github.com/ExpDev07/coronavirus-tracker-api/blob/master/app/utils/countries.py) dictionary using `MS Zaandam` and/or `Diamond Princess` as a country name. And it returned `XX` with a warning since no such countries exist. And such warnings appear only once per hour since the responses are cached for exactly one hour. No magic here, nothing to worry about.

Among other reasons, the warnings appear, if the JHU CSEE misspells or abbreviates a country name or uses some non ISO standard country name in their csv files. In this situation a new alias must be added to the dictionary, programmatically correcting such a mistake.

The case of `MS Zaandam` and `Diamond Princess` is a bit tricky - these are proper names of real ships. I've been thinking about suppressing such warnings in these two particular cases, but(!) I decided not to do so, since ships are "tricky". They can move from port to port, i.e. from country to country or disappear, i.e. move to international waters, or the passengers can leave the ship, thus decreasing(!) the number of confirmed cases, deaths etc. So I prefer to keep these warnings around as a reminder of this trickiness.

**Comment by Toxyl** [comment]
> And such warnings appear only once per hour since the responses are cached for exactly one hour. No magic here, nothing to worry about.

I know that. That wasn't the point. The point being that they appear *multiple* times in *one* update.

**Comment by Kilo59** [comment]
@Toxyl @Bost 
The log message appearing multiple times isn't necessarily an issue either. You could be seeing the logs from multiple processes.
The application is deployed using the [gunicorn](https://gunicorn.org/) which uses multiple workers.
http://docs.gunicorn.org/en/latest/settings.html#worker-processes

I believe the number of workers  (if not set) is determined by the number of cores on the host machine.
These are independent processes that will each have their own cache.
Gunicorn orchestrates them. 
https://github.com/ExpDev07/coronavirus-tracker-api/blob/master/Procfile

**Comment by Kilo59** [comment]
@Toxyl @Bost 
Added more logging in #290
